packages/@atjson/peritext/src/index.ts

```diff
@@ -0,0 +1,585 @@
+import Document, {
+  Block,
+  Mark,
+  EdgeBehaviour,
+  JSON,
+  AnnotationConstructor,
+  AttributesOf,
+  Annotation,
+  SliceAnnotation,
+  TextAnnotation,
+} from "@atjson/document";
+import uuid from "uuid-random";
+
+export type Peritext = {
+  text: string;
+  blocks: Block[];
+  marks: Mark[];
+};
+
+type Range = `${"[" | "("}${number}..${number}${"]" | ")"}`;
+
+function areParentsEqual(left: string[], right: string[]) {
+  return (
+    left.length === right.length &&
+    left.every((value, index) => value === right[index])
+  );
+}
+
+function attributesWithStableIds(
+  attribute: any,
+  ids: Map<string, string>
+): NonNullable<any> {
+  if (attribute == null) {
+    return null;
+  } else if (Array.isArray(attribute)) {
+    const copy: any[] = [];
+    for (let i = 0, len = attribute.length; i < len; i++) {
+      copy[i] = attributesWithStableIds(attribute[i], ids);
+    }
+    return copy;
+  } else if (attribute instanceof Document) {
+    return attribute;
+  } else if (typeof attribute === "object") {
+    const copy: NonNullable<any> = {};
+    for (const key in attribute) {
+      if (attribute[key] !== undefined) {
+        copy[key] = attributesWithStableIds(attribute[key], ids);
+      }
+    }
+    return copy;
+  } else if (typeof attribute === "string" && ids.has(attribute)) {
```

Comment:
slightly worried about unintended collisions here, but I guess that is extremely unlikely since unstable IDs are UUIDs and stable ids use a reserved construction (`M00000001`)